### PR TITLE
Fix/bug google login

### PR DESCRIPTION
**Pull request by yeontannie**

### CODE REVIEW

File: blog-investigation/src/shared/components/ThemeToggler.jsx [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import React from 'react'
+import {Switch} from 'antd'
+
+function ThemeToggler({theme, toggleTheme}) {
```
Comment (denysmatsevych) [inline blog-investigation/src/shared/components/ThemeToggler.jsx]: remove props

File: blog-investigation/src/shared/components/ThemeToggler.jsx [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+  return (
```
Comment (denysmatsevych) [inline blog-investigation/src/shared/components/ThemeToggler.jsx:8]: use context to get theme, toggleTheme

File: blog-investigation/src/shared/hooks/useAuthService.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    gapi.load("client:auth2", start)
+
+
+    const onSuccessLogin = () => {
+        const token = gapi.auth.getToken().access_token
+        setAccessToken(token)
+    }
+
+    const onFailureLogin = (res) => {
+        console.log('login failed', res)
```
Comment (denysmatsevych) [inline blog-investigation/src/shared/hooks/useAuthService.js:23]: setAccessToken('')

File: blog-investigation/src/shared/hooks/useAuthService.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
+import { gapi } from 'gapi-script'
+import { useUserSettingsContext } from '../store/UserSettingsProvider'
+import { CLIENT_ID } from '../../secret.env'
+
+export const useAuthService = () => {
+    const {setAccessToken} = useUserSettingsContext()
```
Comment (denysmatsevych) [inline blog-investigation/src/shared/hooks/useAuthService.js]: use localStorage instead

File: blog-investigation/src/shared/navigation/Navbar.jsx [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  return (
+    <Header className={theme} style={{width: '100%'}}>
+      <Menu theme={theme} mode="horizontal" defaultSelectedKeys={['1']}>
+        <Menu.Item key="1" className='logo'>
+          <Link to="/">
+            <img src={logo} alt='logo' style={{height: '50px'}}/>
+          </Link>
+        </Menu.Item> 
+        <Menu.Item className='toggler' key="2">
+          <ThemeToggler theme={theme} toggleTheme={toggleTheme} />
```
Comment (denysmatsevych) [inline blog-investigation/src/shared/navigation/Navbar.jsx]: remove props

File: blog-investigation/src/shared/store/UserSettingsProvider.jsx [hunk 6]
```diff
@@ -0,0 +1,37 @@
+import React, { useReducer, useContext, useMemo } from 'react'
+import { UserSettingsContext } from './context'
+import userSettingsReducer, { userSettingsInitialState } from './reducer'
+import { toggleTheme, toggleLanguage, setAccessToken } from './actions'
+
+const UserSettingsProvider = ({children}) => {
+    const [state, dispatch] = useReducer(userSettingsReducer, userSettingsInitialState) 
+    
+    const setTheme = (theme) => {
+        const newTheme = theme ? 'dark' : 'light'
```
Comment (denysmatsevych) [inline blog-investigation/src/shared/store/UserSettingsProvider.jsx]: move to reducer please

File: blog-investigation/src/shared/store/UserSettingsProvider.jsx [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+        localStorage.setItem('theme', newTheme)
+        dispatch(toggleTheme(newTheme))
+    }
+
+    const setLanguage = (language) => {
+        const newLanguage = language === 'uk' ? 'en' : 'uk'
```
Comment (denysmatsevych) [inline blog-investigation/src/shared/store/UserSettingsProvider.jsx]: move to reducer please

File: blog-investigation/src/shared/store/actions.js [hunk 8]
```diff
@@ -0,0 +1,27 @@
+
+export const toggleTheme = (newTheme) => {
+    return {
+        type:"TOGGLE_THEME",
+        payload:{
+            theme:newTheme
```
Comment (denysmatsevych) [inline blog-investigation/src/shared/store/actions.js]: not needed in case of toggling

File: blog-investigation/src/shared/store/actions.js [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+        }
+    }
+}
+
+export const toggleLanguage = (newLanguage) => {
+    return {
+        type:'TOGGLE_LANGUAGE',
+        payload:{
+            language:newLanguage
```
Comment (denysmatsevych) [inline blog-investigation/src/shared/store/actions.js]: not needed in case of toggling

File: blog-investigation/src/shared/store/actions.js [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+        }
+    }
+}
+
+export const setAccessToken = (token) => {
```
Comment (denysmatsevych) [inline blog-investigation/src/shared/store/actions.js]: remove please

File: blog-investigation/src/App.js [hunk 11]
```diff
@@ -1,21 +1,22 @@
...
-        subTitle="2017182818828182881 packages have been installed, good luck."
-        extra={[
-          <Button type="primary" key="console">
-            Thanks
-          </Button>,
-          <Button key="buy">Shoosh...</Button>,
-        ]}
-      />
+      <UserSettingsProvider>
+        <BrowserRouter>
```
Comment (denysmatsevych) [inline blog-investigation/src/App.js]: BrowserRouter first, then UserSettingsProvider

File: blog-investigation/src/index.css [hunk 12]
```diff
@@ -11,3 +11,68 @@ code {
...
+  margin-right: auto;
+}
+
+.ant-menu-overflow-item.ant-menu-item.ant-menu-item-only-child.menu-login {
+  margin-right: 50px;
+}
+
+.ant-layout-content.dark, .ant-layout-content.light {
+  padding: 40px 20px;
+  height: 89.9vh;
```
Comment (denysmatsevych) [inline blog-investigation/src/index.css]: why not 89.7vh? I think 89.8 should be better :))

File: blog-investigation/src/shared/authentication/Login.jsx [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+  useEffect(() => {
+    authService.initClient()
+  }, [])
+
+  return (
+    <span>
+      <GoogleLogin 
+        render={renderProps => (  
+          <Button className='login-btn' onClick={renderProps.onClick} disabled={renderProps.disabled}>
+            <GoogleOutlined/> Login
```
Comment (denysmatsevych) [inline blog-investigation/src/shared/authentication/Login.jsx]: wrap Login with span

File: blog-investigation/src/shared/authentication/Logout.jsx [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+
+function Logout(){
+    const {onSuccessLogout} = useGapiAuth()
+
+    return(
+        <span>
+            <GoogleLogout 
+                render={renderProps => (
+                    <Button className='logout-btn' onClick={renderProps.onClick} disabled={renderProps.disabled}>
+                        <GoogleOutlined/> Logout
```
Comment (denysmatsevych) [inline blog-investigation/src/shared/authentication/Logout.jsx]: wrap Logout with span

File: blog-investigation/src/shared/hooks/useGapiAuth.js [hunk 15]
```diff
@@ -0,0 +1,21 @@
+import { useUserSettingsContext } from "../store/UserSettingsProvider"
+
+export const useGapiAuth = () => {
+    const {setIsLoggedIn} = useUserSettingsContext()
+
+    const onSuccessLogin = (res) => {
+        setIsLoggedIn()
```
Comment (denysmatsevych) [inline blog-investigation/src/shared/hooks/useGapiAuth.js]: pass `true`

File: blog-investigation/src/shared/hooks/useGapiAuth.js [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+        localStorage.setItem('token', res.accessToken)
+    }
+    
+    const onFailureLogin = () => {
+        localStorage.setItem('token', '')
+    }
+    
+    const onSuccessLogout = () => {
+        setIsLoggedIn()
```
Comment (denysmatsevych) [inline blog-investigation/src/shared/hooks/useGapiAuth.js]: pass `false`

File: blog-investigation/src/shared/services/authService.js [hunk 17]
```diff
@@ -0,0 +1,13 @@
+import { gapi } from 'gapi-script'
+import { CLIENT_ID } from '../../secret.env'
+
+export default class authService {
```
Comment (denysmatsevych) [inline blog-investigation/src/shared/services/authService.js:4]: AuthService

File: blog-investigation/src/shared/store/reducer.js [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+            
+            return {
+                ...state,
+                language: payload.language
+            }
+
+        case "SET_IS_LOGGED_IN":
+        return {
+            ...state,
+            isLoggedIn:!payload.isLoggedIn
```
Comment (denysmatsevych) [inline blog-investigation/src/shared/store/reducer.js]: danger behavior, set value from payload

File: blog-investigation/src/shared/store/UserSettingsProvider.jsx [hunk 19]
```diff
@@ -0,0 +1,24 @@
+import React, { useReducer, useContext, useMemo } from 'react'
+import { UserSettingsContext } from './context'
+import userSettingsReducer, { userSettingsInitialState } from './reducer'
+
+const UserSettingsProvider = ({children}) => {
+    const [state, dispatch] = useReducer(userSettingsReducer, userSettingsInitialState) 
+
+    const userSettingsContextValue = useMemo(() => {
+        return {
+            ...state,
```
Comment (denysmatsevych) [inline blog-investigation/src/shared/store/UserSettingsProvider.jsx]: causes rerenders, put theme, language, isLoggedIn as  keys

File: blog-investigation/src/shared/store/UserSettingsProvider.jsx [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+            toggleTheme: (isChecked) => dispatch({type:'TOGGLE_THEME', payload:{isChecked:isChecked}}),
```
Comment (denysmatsevych) [inline blog-investigation/src/shared/store/UserSettingsProvider.jsx]: payload: { isChecked }

File: blog-investigation/src/shared/store/UserSettingsProvider.jsx [hunk 21]
```diff
@@ -0,0 +1,24 @@
...
+            toggleLanguage: (language) => dispatch({type:'TOGGLE_LANGUAGE', payload:{language:language}}),
```
Comment (denysmatsevych) [inline blog-investigation/src/shared/store/UserSettingsProvider.jsx]: payload: { language }

File: blog-investigation/src/shared/store/UserSettingsProvider.jsx [hunk 22]
```diff
@@ -0,0 +1,24 @@
...
+            setIsLoggedIn: () => dispatch({type:"SET_IS_LOGGED_IN", payload:{isLoggedIn:state.isLoggedIn}})
```
Comment (denysmatsevych) [inline blog-investigation/src/shared/store/UserSettingsProvider.jsx]: pass payload

File: blog-investigation/src/shared/store/UserSettingsProvider.jsx (hunk 19 above)
Comment (yeontannie) [reply blog-investigation/src/shared/store/UserSettingsProvider.jsx]: I have state as key, should I add its props again?

Comment (denysmatsevych) [reply blog-investigation/src/shared/store/UserSettingsProvider.jsx]: yes, please add. spread operator `...` generates a new reference

**Review by denysmatsevych — APPROVED** [review]

**Pull request merged**